src/pytorch/pme/pmeCUDA.cu

```diff
@@ -0,0 +1,447 @@
+using namespace torch::autograd;
+using torch::Tensor;
+using torch::TensorOptions;
+using torch::Scalar;
+
+#define CHECK_RESULT(result) \
+    if (result != cudaSuccess) { \
+        throw runtime_error(string("Encountered error ")+cudaGetErrorName(result)+" at "+__FILE__+":"+to_string(__LINE__));\
+    }
+
+static int getMaxBlocks() {
+    // Get an upper limit on how many thread blocks we try to launch based on the size of the GPU.
+
+    int device, numMultiprocessors;
+    CHECK_RESULT(cudaGetDevice(&device));
+    CHECK_RESULT(cudaDeviceGetAttribute(&numMultiprocessors, cudaDevAttrMultiProcessorCount, device));
+    return numMultiprocessors*4;
+}
+
+__global__ void computeDirect(const Accessor<float, 2> pos, const Accessor<float, 1> charge, Accessor<int, 2> neighbors, const Accessor<float, 2> deltas,
+                              const Accessor<float, 1> distances, const Accessor<int, 2> exclusions, Accessor<float, 2> posDeriv,
+                              Accessor<float, 1> chargeDeriv, Accessor<float, 1> energyBuffer, float alpha, float coulomb) {
+    int numAtoms = pos.size(0);
+    int numNeighbors = neighbors.size(1);
+    int maxExclusions = exclusions.size(1);
+    double energy = 0;
+
+    for (int index = blockIdx.x*blockDim.x+threadIdx.x; index < numNeighbors; index += blockDim.x*gridDim.x) {
+        int atom1 = neighbors[0][index];
+        int atom2 = neighbors[1][index];
+        float r = distances[index];
+        bool include = (atom1 > -1);
+        for (int j = 0; include && j < maxExclusions && exclusions[atom1][j] >= atom2; j++)
+            if (exclusions[atom1][j] == atom2)
+                include = false;
+        if (include) {
+            float invR = 1/r;
+            float alphaR = alpha*r;
+            float expAlphaRSqr = expf(-alphaR*alphaR);
+            float erfcAlphaR = erfcf(alphaR);
+            float prefactor = coulomb*invR;
+            float c1 = charge[atom1];
+            float c2 = charge[atom2];
+            energy += prefactor*erfcAlphaR*c1*c2;
+            atomicAdd(&chargeDeriv[atom1], prefactor*erfcAlphaR*c2);
+            atomicAdd(&chargeDeriv[atom2], prefactor*erfcAlphaR*c1);
+            float dEdR = prefactor*c1*c2*(erfcAlphaR+alphaR*expAlphaRSqr*M_2_SQRTPI)*invR*invR;
+            for (int j = 0; j < 3; j++) {
+                atomicAdd(&posDeriv[atom1][j], -dEdR*deltas[index][j]);
+                atomicAdd(&posDeriv[atom2][j], dEdR*deltas[index][j]);
+            }
+        }
+    }
+
+    // Subtract excluded interactions to compensate for the part that is
+    // incorrectly added in reciprocal space.
+
+    float dr[3];
+    for (int index = blockIdx.x*blockDim.x+threadIdx.x; index < numAtoms*maxExclusions; index += blockDim.x*gridDim.x) {
+        int atom1 = index/maxExclusions;
+        int atom2 = exclusions[atom1][index-atom1*maxExclusions];
+        if (atom2 > atom1) {
+            for (int j = 0; j < 3; j++)
+                dr[j] = pos[atom1][j]-pos[atom2][j];
```

Comment:
Dont you have to apply PBC here?

Reply:
Exclusions apply only to a single copy of each other atom, the one in the same periodic copy.  See the discussion of this in the class documentation.